<!--
	@file algorithms.html
	@brief the AngularJS partial 'algorithms'
-->

<h1>DnaAssembler - {{'ALGORITHMS' | translate}}</h1><br>

	<h3>{{'APPLICATION_PARAMETERS' | translate}}:</h3>
	<ul>
		<li id="assembling_alg"><b>{{'ASSEMBLING' | translate}}:</b></li><br>
		<ul>
			<li id="k_alg"><b>{{'K' | translate}}: </b><em id="k_inst">{{'K_INSTRUCTION' | translate}}</em></li><br>
			<li id="genome_length_alg"><b>{{'GENOME_LENGTH' | translate}}: </b><em id="genome_length_inst">{{'GENOME_LENGTH_INSTRUCTION' | translate}}</em></li><br>
			<li id="insert_size_mean_inward_alg"><b>{{'INSERT_SIZE_MEAN_INWARD' | translate}}: </b><em id="insert_size_mean_inward_inst">{{'INSERT_SIZE_MEAN_INWARD_INSTRUCTION' | translate}}</em></li><br>
			<li id="insert_size_std_dev_inward_alg"><b>{{'INSERT_SIZE_STD_DEV_INWARD' | translate}}: </b><em id="insert_size_std_dev_inward_inst">{{'INSERT_SIZE_STD_DEV_INWARD_INSTRUCTION' | translate}}</em></li><br>
			<li id="insert_size_mean_outward_alg"><b>{{'INSERT_SIZE_MEAN_OUTWARD' | translate}}: </b><em id="insert_size_mean_outward_inst">{{'INSERT_SIZE_MEAN_OUTWARD_INSTRUCTION' | translate}}</em></li><br>
			<li id="insert_size_std_dev_outward_alg"><b>{{'INSERT_SIZE_STD_DEV_OUTWARD' | translate}}: </b><em id="insert_size_std_dev_outward_inst">{{'INSERT_SIZE_STD_DEV_OUTWARD_INSTRUCTION' | translate}}</em></li><br>
			<li id="quality_threshold_alg"><b>{{'QUALITY_THRESHOLD' | translate}}: </b><em id="quality_threshold_inst">{{'QUALITY_THRESHOLD_INSTRUCTION' | translate}}</em></li><br>
			<li id="bfcounter_threshold_alg"><b>{{'BFCOUNTER_THRESHOLD' | translate}}: </b><em id="bfcounter_threshold_inst">{{'BFCOUNTER_THRESHOLD_INSTRUCTION' | translate}}</em></li><br>
			<li id="single_edge_counter_threshold_alg"><b>{{'SINGLE_EDGE_COUNTER_THRESHOLD' | translate}}: </b><em id="single_edge_counter_threshold_inst">{{'SINGLE_EDGE_COUNTER_THRESHOLD_INSTRUCTION' | translate}}</em></li><br>
			<li id="paired_reads_pet_threshold_from_alg"><b>{{'PAIRED_READS_PET_THRESHOLD_FROM' | translate}}: </b><em id="paired_reads_pet_threshold_from_inst">{{'PAIRED_READS_PET_THRESHOLD_FROM_INSTRUCTION' | translate}}</em></li><br>
			<li id="paired_reads_pet_threshold_to_alg"><b>{{'PAIRED_READS_PET_THRESHOLD_TO' | translate}}: </b><em id="paired_reads_pet_threshold_to_inst">{{'PAIRED_READS_PET_THRESHOLD_TO_INSTRUCTION' | translate}}</em></li><br>
			<li id="paired_reads_mp_threshold_from_alg"><b>{{'PAIRED_READS_MP_THRESHOLD_FROM' | translate}}: </b><em id="paired_reads_mp_threshold_from_inst">{{'PAIRED_READS_MP_THRESHOLD_FROM_INSTRUCTION' | translate}}</em></li><br>
			<li id="paired_reads_mp_threshold_to_alg"><b>{{'PAIRED_READS_MP_THRESHOLD_TO' | translate}}: </b><em id="paired_reads_mp_threshold_to_inst">{{'PAIRED_READS_MP_THRESHOLD_TO_INSTRUCTION' | translate}}</em></li><br>
			<li id="correction_alg"><b>{{'CORRECTION' | translate}}: </b><em id="correction_inst">{{'CORRECTION_INSTRUCTION' | translate}}</em></li><br>
			<li id="paired_reads_algorithm_alg"><b>{{'PAIRED_READS_ALGORITHM' | translate}}: </b><em id="paired_reads_algorithm_inst">{{'PAIRED_READS_ALGORITHM_INSTRUCTION' | translate}}</em></li><br>
			<li id="input_file_i1_1_alg"><b>{{'INPUT_FILE_I1_1' | translate}}: </b><em id="input_file_i1_1_inst">{{'INPUT_FILE_I1_1_INSTRUCTION' | translate}}</em></li><br>
			<li id="input_file_i1_2_alg"><b>{{'INPUT_FILE_I1_2' | translate}}: </b><em id="input_file_i1_2_inst">{{'INPUT_FILE_I1_2_INSTRUCTION' | translate}}</em></li><br>
			<li id="input_file_o1_1_alg"><b>{{'INPUT_FILE_O1_1' | translate}}: </b><em id="input_file_o1_1_inst">{{'INPUT_FILE_O1_1_INSTRUCTION' | translate}}</em></li><br>
			<li id="input_file_o1_2_alg"><b>{{'INPUT_FILE_O1_2' | translate}}: </b><em id="input_file_o1_2_inst">{{'INPUT_FILE_O1_2_INSTRUCTION' | translate}}</em></li><br>
		</ul>
		<li id="alignment_alg"><b>{{'ALIGNMENT' | translate}}:</b></li><br>
		<ul>
			<li id="gap_penalty_alg"><b>{{'GAP_PENALTY' | translate}}: </b><em id="gap_penalty_inst">{{'GAP_PENALTY_INSTRUCTION' | translate}}</em></li><br>
			<li id="similarity_matrix_alg"><b>{{'SIMILARITY_MATRIX' | translate}}: </b><em id="similarity_matrix_inst">{{'SIMILARITY_MATRIX_INSTRUCTION' | translate}}</em></li><br>
			<li id="align_algorithm_alg"><b>{{'ALIGN_ALGORITHM' | translate}}: </b><em id="align_algorithm_inst">{{'ALIGN_ALGORITHM_INSTRUCTION' | translate}}</em></li><br>
			<li id="circular_alg"><b>{{'CIRCULAR' | translate}}: </b><em id="circular_inst">{{'CIRCULAR_INSTRUCTION' | translate}}</em></li><br>
			<li id="sequence_file_alg"><b>{{'SEQUENCE_FILE' | translate}}: </b><em id="sequence_file_inst">{{'SEQUENCE_FILE_INSTRUCTION' | translate}}</em></li><br>
			<li id="reference_file_alg"><b>{{'REFERENCE_FILE' | translate}}: </b><em id="reference_file_inst">{{'REFERENCE_FILE_INSTRUCTION' | translate}}</em></li><br>
		</ul>
	</ul>

	<h3>{{'ARTICLES_LIST' | translate}}</h3>
	<ul>
		<li id="main_algorithms_article">Nowak R. Assembly of repetitive regions using next-generation sequencing data. Biocybern Biomed Eng (2015).</li><br>
		<li id="velvet_paired_reads_article">Zerbino DR, Birney E. Velvet: Algorithms for de novo short read assembly using de Bruijn graphs. Genome Research. 2008;18(5):821-829. doi:10.1101/gr.074492.107.</li><br>
		<li id="needleman_wunsch_article">Needleman, Saul B.; and Wunsch, Christian D. (1970). "A general method applicable to the search for similarities in the amino acid sequence of two proteins". Journal of Molecular Biology 48 (3): 443â€“53.</li><br>
		<li id="hirschberg_article">D. S. Hirschberg. A linear space algorithm for computing maximal common subsequences. Comm. A.C.M. 18(6) p341-343, 1975.</li><br><br>
	</ul>
